agasc/agasc.py

```diff
@@ -133,6 +147,153 @@ def get_ra_decs(agasc_file):
+    """
+    if cache:
+        if isinstance(h5_file, tables.file.File):
+            h5_file = h5_file.filename
+        data = _read_h5_table_cached(h5_file, path)
+        out = data[row0:row1]
+    else:
+        out = _read_h5_table(h5_file, path, row0, row1)
+
+    return out
+
+
+@functools.lru_cache(maxsize=1)
+def _read_h5_table_cached(
+    h5_file: str | Path,
+    path: str,
+) -> np.ndarray:
+    return _read_h5_table(h5_file, path, row0=None, row1=None)
+
+
+def _read_h5_table(
+        h5_file: str | Path | tables.file.File,
+        path: str,
+        row0: None | int,
+        row1: None | int,
+    ) -> np.ndarray:
+    if isinstance(h5_file, tables.file.File):
+        out = _read_h5_table_from_open_h5_file(h5_file, path, row0, row1)
+    else:
+        with tables.open_file(h5_file) as h5:
+            out = _read_h5_table_from_open_h5_file(h5, path, row0, row1)
+
+    out = np.asarray(out)  # Convert to structured ndarray (not recarray)
+    return out
+
+def _read_h5_table_from_open_h5_file(h5, path, row0, row1):
+    data = getattr(h5.root, path)
+    out = data.read(start=row0, stop=row1)
+    return out
+
+
+def get_agasc_filename(agasc_file: Optional[str | Path]=None):
+    """Get a matching AGASC file name from ``agasc_file``.
+
+    If ``agasc_file`` is None the return value is (in order of precedence):
+    - ``${AGASC_HDF5_FILE}`` if that environment variable is defined.
+    - ``${SKA}/data/agasc/miniagasc<version>.h5`` where ``<version>`` is the latest
+      version found in the ``${SKA}/data/agasc`` directory. This is the "full" AGASC.
+
+    If ``agasc_file`` ends in ``.h5`` the return value is ``agasc_file``.
+
+    Other values of ``agasc_file`` are interpreted as a root name for files within
+    the default AGASC directory. The return value is the latest version of files
+    matching that root name. For example, for ``agasc_file="miniagasc"`` then the
+    return value is ``${SKA}/data/agasc/miniagasc_<version>.h5`` where ``<version>``
+    is the latest version found in the ``${SKA}/data/agasc`` directory.
+
+    File versions are required to match the pattern ``1p[0-9]+``. Note that release
+    candidate files (e.g. "agasc1p8rc3.h5") are not included, only full releases are
+    matched.
+
+    The default AGASC directory is the environment variable ``${AGASC_DIR}`` if defined,
+    otherwise ``${SKA}/data/agasc``.
+
+    Parameters
+    ----------
+    agasc_file : str, Path, optional
+        AGASC file name (default=None)
+
+    Returns
+    -------
+    filename : str
+        Matching AGASC file name
+
+    Raises
+    ------
+    FileNotFoundError
+        If no matching AGASC file is found.
+    """
+    if agasc_file is None:
+        agasc_file = os.environ.get('AGASC_HDF5_FILE')
+        if agasc_file is None:
+            return get_agasc_filename("miniagasc")
+
+    agasc_file = str(agasc_file)
+
+    if agasc_file.endswith('.h5'):
+        out = agasc_file
```

Comment:
This logic is inconsistent with the current MATLAB implementation.  For the MATLAB tools, the environment variable `AGASC_HDF5_FILE` only ever specifies the filename to be used.  It is assumed that this file will always live in `${SKA}/data/agasc` or `${SKA_DATA_DIR_FOR_MATLAB_TOOLS}/agasc`, if the `SKA_DATA_DIR_FOR_MATLAB_TOOLS` environment variable exists.

It looks like this code is expecting this environment variable to contain the full path to the agasc file if it ends with `.h5`, based on the check on line 291, which checks if the path exists, but since it's just the file name, that doesn't exist in whatever the pwd is when this code is run.

Reply:
Ok I'll adjust the code accordingly.

Reply:
Though actually, what does make sense for the most reasonable implementation of AGASC_HDF5_FILE?  Should it just be a full path?  Since we're looking at putting the "change the file" into a release it does seem worth taking a moment to figure out if there are advantages to making a change.

Reply:
We can implement is however.  My understanding was that `AGASC_HDF5_FILE` was going to name the file to use, and `AGASC_DIR` or `SKA/data/agasc` was going to be the location of said file.  To be fair, the MATLAB tools are not currently using the `AGASC_DIR` environment variable though, so I may update the MATLAB code to use it too.